File: ios/BleNative/Utils/ObjectIdGenerator.swift
import Foundation
import CoreBluetooth

enum ObjectIdGenerators {
    static let peripherals = ObjectIdGenerator<CBPeripheral>()
    static let services = ObjectIdGenerator<CBService>()
    static let characteristics = ObjectIdGenerator<CBCharacteristic>()
}

class ObjectIdGenerator<Element: Hashable> {         
    private var objectToIdentifier: [Element: Int32] = [:]
    
    private var currentId: Int32 = 0
    
    fileprivate init() {}
    
    func id(for element: Element) -> Int32 {
        if let identifier = objectToIdentifier[element] {
            return identifier
        }
        currentId += 1
        objectToIdentifier[element] = currentId
        return currentId
    }
}

File: android/src/main/java/com/polidea/blenative/utils/ObjectIdGenerator.kt
package com.polidea.blenative.utils

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattService

object ObjectIdGenerators {
    val devices = ObjectIdGenerator<BluetoothDevice>()
    val services = ObjectIdGenerator<BluetoothGattService>()
    val characteristics = ObjectIdGenerator<BluetoothGattCharacteristic>()
}

class ObjectIdGenerator<Element> {
    private var objectToIdentifier = mutableMapOf<Element, Int>()

    private var currentId = 0

    fun idForElement(element: Element): Int {
        val identifier = objectToIdentifier[element]
        if (identifier != null) {
            return identifier
        }
        objectToIdentifier[element] = ++currentId
        return currentId
    }
}